bots/github-push

```diff
@@ -0,0 +1,125 @@
+# cached in the Git config file to speed up subsequent invocations.
+
+import subprocess
+import argparse
+import sys
+import time
+
+from task import github
+
+def execute(*args):
+    try:
+        output = subprocess.check_output(args, universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        sys.exit(ex.returncode)
+    return output
+
+def git(*args):
+    return execute("git", *args)
+
+def git_config(name):
+    try:
+        output = subprocess.check_output([ "git", "config", name ], universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        return None
+    return output.strip()
+
+def main():
+    parser = argparse.ArgumentParser(description='Force push after a rewrite')
+    parser.add_argument('--repo', help="The GitHub repository to work with", default="cockpit-project/cockpit")
+    opts = parser.parse_args()
+
+    api = github.GitHub(repo=opts.repo)
+
+    local = git("rev-parse", "HEAD").strip()
+    remote = git("rev-parse", "@{u}").strip()
+
+    if local == remote:
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    ( tracking_remote, tracking_ref ) = git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}").strip().split("/")
+    tracking_url = git("remote", "get-url", tracking_remote).strip()
+
+    push_dest = tracking_remote
+    push_refspecs = local + ":" + tracking_ref
+    push_lease = tracking_ref + ":" + remote
+
+    pull_number = None
+    pull_config = git_config("pull.%s.%s" % (tracking_remote, tracking_ref))
+    if not pull_config:
+        # Find the pull request that matches the tracking branch
+        pulls = api.pulls()
+        for p in pulls:
+            ref = p["head"]["ref"]
+            url = "git@github.com:%s.git" % p["head"]["repo"]["full_name"]
+            if ref == tracking_ref and url == tracking_url:
+                pull_number = p["number"]
+                break
+        if not pull_number:
+            sys.stderr.write("No pull request found for %s/%s\n" % (tracking_remote, tracking_ref))
+            return 1
+        git("config", "pull.%s.%s" % (tracking_remote, tracking_ref), str(pull_number))
+    else:
+        pull_number=int(pull_config)
+
+    diff = git("diff",remote, local)
+    old_commits_link = "https://github.com/" + opts.repo + "/commits/" + remote
+    diff_link = "https://github.com/" + opts.repo + "/compare/" + remote + ".." + local
+
+    def make_comment(title, include_old_commits=False):
+        diff_lines = diff.count("\n")
+        if include_old_commits:
+            first_line = "[Old commits](%s) | [Diff](%s)" % (old_commits_link, diff_link)
+        else:
+            first_line = "[Diff](%s)" % (diff_link)
+        if diff_lines < 500:
+            diff_open = "open" if diff_lines < 50 else ""
+            return ("<details %s><summary>%s</summary>\n\n%s\n```diff\n%s```\n</details>\n"
+                    % (diff_open, title, first_line, diff))
```

Comment:
Is this magic GitHub syntax to create an expander in a comment? Nice!

Reply:
It's actually non-magic HTML... :-)